include/zephyr/sys/ref.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2023 Michael Zimmermann
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_REF_H_
+#define ZEPHYR_INCLUDE_SYS_REF_H_
+
+#include <limits.h>
+#include <stdbool.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup sys_ref_apis Reference count APIs
+ * @{
+ */
+
+/* atomics are usually `long`. If they are larger, that's fine because
+ * we'll just artificially limit them . If they were smaller, we'd be unable
+ * to check if an overflow happened. This wouldn't be necessary if the atomic
+ * API told us the max value of an atomic.
+ */
+BUILD_ASSERT(sizeof(atomic_val_t) >= sizeof(long));
+
+/**
+ * @brief An atomic reference count.
+ *
+ * The main use case is to share the same data between several users and free
+ * it when all references were released. This can also be used on SMP systems.
+ */
+struct sys_ref {
+	atomic_t refs;
+};
+
+/** @brief Initialize sys_ref at compile-time. */
+#define SYS_REF_INIT()                                                                             \
+	{                                                                                          \
+		.refs = ATOMIC_INIT(1),                                                            \
+	}
+
+/** @brief Initialize sys_ref at runtime. */
+static inline void sys_ref_init(struct sys_ref *const ref)
+{
+	*ref = (struct sys_ref){
+		.refs = ATOMIC_INIT(1),
+	};
+}
+
+/**
+ * @brief Create a new reference to `ref`.
+ *
+ * @return The new reference.
+ */
+static inline struct sys_ref *sys_ref(struct sys_ref *const ref)
+{
+	const atomic_val_t old_rc = atomic_inc(&ref->refs);
+
+	__ASSERT_NO_MSG(old_rc <= LONG_MAX);
```

Comment:
I don't think an assertion is sufficient here.  This is a runtime error, it's entirely possible for code to "correctly" use this framework, disable assertions in production code, and be subject to overflow issues that aren't detected (even worse: security vulnerabilities due to e.g. attacker-controlled premature destruction of the referenced object!).  Semaphores have essentially the same requirements for the same reason, and expose the limit as part of the API.

This either needs to detect the overflow and return an error condition, or else return the actual value (and document the limits) so that calling code can do the right thing on its own.

Reply:
oops :see_no_evil: 
I'll have to think about how to handle this properly. I mean replacing the assertion with a check that returns NULL in this case is probably the right way to do this. Neither net_buf_ref nor bt_conn_ref handle this currently. And none of their users check the return value because they simply assume those APIs must always succeed.

Not having this as an assertion also needs more careful handling. We'll have to do this check before calling atomic_inc. Otherwise we'll just return an error for caller1 and caller1 will then see a count of 0 and happily increase it to 1.

Reply:
Yes, the idea is to emphasize that the reference count should (ideally) always reflect how many pointer variables you have pointing at the same object. That makes it natural to return a new pointer so that you can do:
```c
new_ref = sys_ref(old_ref);
```

In the same spirit, it'd be nice if the `unref()` would be able take a pointer to a pointer and set the old reference to NULL after decrementing the count, however that won't be possible with this kind of design where you embed the generic ref struct as part of a parent struct.

Reply:
> however that won't be possible with this kind of design where you embed the generic ref struct as part of a parent struct

...unless we do generator macros that output type-specific reference count functions